client/client.go

```diff
@@ -2174,33 +2179,76 @@ func (c *Client) allocSync() {
 			}
 			c.allocLock.RUnlock()
 
-			// Successfully updated allocs, reset map and ticker.
-			// Always reset ticker to give loop time to receive
-			// alloc updates. If the RPC took the ticker interval
-			// we may call it in a tight loop before draining
-			// buffered updates.
-			updates = make(map[string]*structs.Allocation, len(updates))
+			// Successfully updated allocs. Reset ticker to give loop time to
+			// receive new alloc updates. Otherwise if the RPC took the ticker
+			// interval we may call it in a tight loop reading empty updates.
+			updateTicks = 0
 			syncTicker.Reset(allocSyncIntv)
 		}
 	}
 }
 
-func (c *Client) filterAcknowledgedUpdates(updates map[string]*structs.Allocation) []*structs.Allocation {
+// updatesToSync returns a list of client allocation updates we need to make in
+// this tick of the allocSync. It returns nil if there's no updates to make
+// yet. The caller is responsible for restoring the c.allocUpdates map if it
+// can't successfully send the updates.
+func (c *Client) updatesToSync(updateTicks int) []*structs.Allocation {
+
+	c.allocUpdatesLock.Lock()
+	defer c.allocUpdatesLock.Unlock()
+
+	// Fast path if there are no pending updates
+	if len(c.allocUpdates) == 0 {
+		return nil
+	}
+
+	// Ensure we never send an update before we've had at least one sync from
+	// the server
+	select {
+	case <-c.serversContactedCh:
+	default:
+		return nil
+	}
+
+	toSync, urgent := c.filterAcknowledgedUpdates(c.allocUpdates)
+
+	// Only update every 5th tick if there's no priority updates
+	if updateTicks%5 != 0 && !urgent {
+		return nil
+	}
+
+	// Clear here so that allocrunners can queue up the next set of updates
+	// while we're waiting to hear from the server
+	c.allocUpdates = make(map[string]*structs.Allocation, len(c.allocUpdates))
+
+	return toSync
+}
+
+// filteredAcknowledgedUpdates returns a list of client alloc updates with the
+// already-acknowledged updates removed, and the highest priority of any update.
```

Comment:
I'd go a step further and suggest extracting the `allocUpdates` map into its own little data structure and let it handle its own locking. There's quite a bit of implementation detail around what it's used for splattered all over these Client functions.

Reply:
> If I understand correctly, the caller must hold at least a read lock on `c.allocUpdatesLock` when calling this function? If so, would it be worth adding a note to the function comment about this for future readers?

Well it's not a `RWMutex`, but yes. Really the only reason this is in its own function at all now is that we need to take the `c.allocLock` to read from the allocrunners and having it in its own function lets us scope-down that lock.

> I'd go a step further and suggest extracting the `allocUpdates` map into its own little data structure and let it handle its own locking. There's quite a bit of implementation detail around what it's used for splattered all over these Client functions.

So in other words, make `updatesToSync` and `filterAcknowledgedUpdates` methods on the `allocUpdates` data structure? We'd need to pass the `*Client` as a parameter so that we can access the allocrunners. But yeah that seems reasonable. Let me take a quick try at that.

Reply:
@shoenig @jrasell I've refactored this by pulling out the `allocUpdates` to their own struct `pendingAllocUpdates` (so as not to conflict with the other struct named `allocUpdates` which is the updates received from the server!). I think this makes the whole locking situation a lot cleaner. Let me know what you think.